plugin/evm/validators_transitioner_test.go

```diff
@@ -3,164 +3,214 @@
+		},
+		{
+			name:         "no steps non-zero epoch",
+			steps:        []Step{},
+			epoch:        1,
+			validatorIDs: validatorIDs,
+		},
+		{
+			name:  "one step epoch before",
+			epoch: 4,
+			steps: []Step{
+				{Epoch: 5, Cutoff: 2},
+			},
+			validatorIDs: validatorIDs,
+		},
+		{
+			name:  "one step epoch exact",
+			epoch: 5,
+			steps: []Step{
+				{Epoch: 5, Cutoff: 2},
+			},
+			validatorIDs: append(validatorIDs[:2], providerIDs[0]),
+		},
+		{
+			name:  "one step epoch after",
+			epoch: 6,
+			steps: []Step{
+				{Epoch: 5, Cutoff: 2},
+			},
+			validatorIDs: append(validatorIDs[:2], providerIDs[0]),
+		},
+		{
+			name:  "two steps epoch before first",
+			epoch: 4,
+			steps: []Step{
+				{Epoch: 5, Cutoff: 2},
+				{Epoch: 10, Cutoff: 1},
+			},
+			validatorIDs: validatorIDs,
+		},
+		{
+			name:  "two steps epoch exact first",
+			epoch: 5,
+			steps: []Step{
+				{Epoch: 5, Cutoff: 2},
+				{Epoch: 10, Cutoff: 1},
+			},
+			validatorIDs: append(validatorIDs[:2], providerIDs[0]),
+		},
+		{
+			name:  "two steps epoch after first",
+			epoch: 6,
+			steps: []Step{
+				{Epoch: 5, Cutoff: 2},
+				{Epoch: 10, Cutoff: 1},
+			},
+			validatorIDs: append(validatorIDs[:2], providerIDs[0]),
+		},
+		{
+			name:  "two steps epoch before second",
+			epoch: 9,
+			steps: []Step{
+				{Epoch: 5, Cutoff: 2},
+				{Epoch: 10, Cutoff: 1},
+			},
+			validatorIDs: append(validatorIDs[:2], providerIDs[0]),
+		},
+		{
+			name:  "two steps epoch exact second",
+			epoch: 10,
+			steps: []Step{
+				{Epoch: 5, Cutoff: 2},
+				{Epoch: 10, Cutoff: 1},
+			},
+			validatorIDs: append(validatorIDs[:1], providerIDs[0], providerIDs[1]),
+		},
+		{
+			name:  "two steps epoch after second",
+			epoch: 11,
+			steps: []Step{
+				{Epoch: 5, Cutoff: 2},
+				{Epoch: 10, Cutoff: 1},
+			},
+			validatorIDs: append(validatorIDs[:1], providerIDs[0], providerIDs[1]),
+		},
+	}
+
+	for _, test := range tests {
+		test := test
+		t.Run(test.name, func(t *testing.T) {
+			t.Parallel()
+
+			v := &ValidatorsDefault{
+				validators: validators,
+				steps:      test.steps,
+			}
+
+			transitioner := NewValidatorsTransitioner(logging.NoLog{}, v, mockProvider, WithStepSize(1))
+
+			got, err := transitioner.ByEpoch(test.epoch)
+			require.NoError(t, err)
+
+			assert.Len(t, got, len(test.validatorIDs))
+			for _, validatorID := range test.validatorIDs {
+				assert.Contains(t, got, validatorID)
+			}
+		})
+	}
+}
+
+func fakeProviders(epoch uint64) map[ids.ShortID]uint64 {
+
+	providers := make(map[ids.ShortID]uint64)
+	for i := 0; i < int(epoch); i++ {
+		providers[ids.ShortID{byte(i)}] = uint64(i)
+	}
+
+	return providers
+}
+
+// fakeValidators gives a different set of validators compared to fakeProviders() to have no overlap for now
+func fakeValidators(epoch uint64) map[ids.ShortID]uint64 {
+
+	providers := make(map[ids.ShortID]uint64)
+	for i := 0; i < int(epoch); i++ {
+		providers[ids.ShortID{byte(i + int(epoch))}] = uint64(i)
```

Comment:
This means on epoch zero, validators and providers will return the same. It also doesn't work for other epochs with higher validator numbers. For example, on epoch 4, validators will be 0, 1, 2, 3, 4, 5, 6, ..., while providers will be 4, 5, 6, 7, ...